cmd/tools/grafana/dashboard_test.go

```diff
@@ -653,59 +653,60 @@ func TestLegends(t *testing.T) {
 func checkLegends(t *testing.T, path string, data []byte) {
 	// collect all legends
 	dashPath := shortPath(path)
-	gjson.GetBytes(data, "panels").ForEach(func(key, value gjson.Result) bool {
+
+	visitAllPanels(data, func(path string, key, value gjson.Result) {
 		doLegends(t, value, dashPath)
-		value.Get("panels").ForEach(func(key2, value2 gjson.Result) bool {
-			doLegends(t, value2, dashPath)
-			return true
-		})
-		return true
 	})
 }
 
 func doLegends(t *testing.T, value gjson.Result, dashPath string) {
-	expectedCals := []string{"mean", "lastNotNull", "max"}
-	expectedDisplayMode := "table"
-	expectedPlacement := "bottom"
+	wantDisplayMode := "table"
+	wantPlacement := "bottom"
 
 	kind := value.Get("type").String()
-	if kind == "row" {
+	if kind == "row" || kind == "piechart" {
 		return
 	}
 	optionsData := value.Get("options")
 	if legendData := optionsData.Get("legend"); legendData.Exists() {
-		legendDisplayeMode := legendData.Get("displayMode").String()
+		legendDisplayMode := legendData.Get("displayMode").String()
 		legendPlacementData := legendData.Get("placement").String()
 		title := value.Get("title").String()
-		if calcsData := legendData.Get("calcs"); calcsData.Exists() {
-			var calcsSlice []string
-			calcsData.ForEach(func(key, val gjson.Result) bool {
-				calcsSlice = append(calcsSlice, val.String())
-				return true
-			})
-			checkCalcs(t, calcsSlice, expectedCals, dashPath, title)
+		calcsData := legendData.Get("calcs").Array()
```

Comment:
few panels don't have `calc` object

Reply:
tested and `test` passes with this.

Reply:
right, it's fine when `calcs` is missing - we get an empty array in that case